Simultaneous training and rollout central lock

alexander-zap opened the pull request.

Commit:
Simultaneous training and rollout

- separate locks for training and rollout
- policies for each agent

Commit:
Made the lock centralized (changed from push to pull principle)

Commit:
removed print

Luke100000 commented on async_gym_agents/agents/on_policy_injector.py:
```diff
@@ -53,28 +56,33 @@ def _episode_generator(self, index: int) -> list[Transition]:
+            # check if training_policy.pt file exists
+            if not os.path.exists("initial_training_policy.pt"):
+                th.save(self.training_policy, f"initial_training_policy.pt")
+            self.rollout_policies[index] = th.load(f"initial_training_policy.pt", weights_only=False)
```
Why here, and not in initialize threads?

alexander-zap replied:
`self.training_policy` might not have been set there.

Luke100000 commented on async_gym_agents/agents/on_policy_injector.py:
```diff
@@ -53,28 +56,33 @@ def _episode_generator(self, index: int) -> list[Transition]:
 
+        with self.training_policy_lock:
+            # check if training_policy.pt file exists
+            if not os.path.exists("initial_training_policy.pt"):
```
```py
    buffer = io.BytesIO()
    th.save(self.training_policy, buffer)
    buffer.seek(0)
    self.rollout_policies[index] = th.load(buffer, weights_only=False)
```

Luke100000 commented on async_gym_agents/agents/off_policy_injector.py:
```diff
@@ -140,14 +141,19 @@ def _episode_generator(self, index: int) -> list[Transition]:
 
+        with self.training_policy_lock:
+            # check if training_policy.pt file exists
+            if not os.path.exists("initial_training_policy.pt"):
```
More than three lines -> custom method :)

Commit:
Refactoring

- temporary model saving to buffer instead of to path
- model copying as function

Commit:
moved model copying to _initialized_threads

Commit:
only update rollout_policy if training_policy was updated

Commit:
removed print

Commit:
Made version check and lock part of super class sync methods
